### ISSUE TITLE
Error: read ECONNRESET when putting file to OpenSSH windows server

### ISSUE THREAD

**Issue by htdan** [theophilusx/ssh2-sftp-client#152]
No problem when putting the same file to Linux SSH server, nor when using Filezilla to upload the same file to the Windows server. 

Tried the following two versions of OpenSSH server on windows server and windows10, same observation.
OpenSSH_for_Windows_7.7p1
OpenSSH_for_Windows_8

Specifically, the error is received when 'end' the client once the file is uploaded. 

The client side code is as follows.
```
sftp.connect(config).then(() => {
    console.log("Uploading file to sftp")  
    var rstream = fs.createReadStream('test.txt');
    return sftp.put(rstream, 'test.txt');
  })
  .then(() => {
    console.log("Shutting down sftp connection")
    return sftp.end();
  })
  .catch((err) => {
    console.log(err, 'catch error');
});
```
The error message:
Uploading file to sftp
Shutting down sftp connection
Error: read ECONNRESET
errors.js:373
    at TCP.onStreamRead (internal/stream_base_commons.js:162:27)
Emitted 'error' event at:
    at Socket.<anonymous> (/Users/hdan/study/javascript/ssh2-sftp-client/node_modules/ssh2/lib/client.js:306:10)
    at Socket.emit (events.js:197:13)
    at emitErrorNT (internal/streams/destroy.js:82:8)
    at emitErrorAndCloseNT (internal/streams/destroy.js:50:3)
    at processTicksAndRejections (internal/process/next_tick.js:76:17)

How we can fix the problem? Thanks

**Comment by theophilusx** [comment]
Unfortunately, I don't have a Windows system and cannot help debug this issue.

This looks like it might be errors on the remote end. At the very least, it is errors being generated by the underlying SSH and sftp-streams library and as these are separate libraries which this module depends on, not a lot can be done at this level. What I would try is

1. Just use the filename rather than a stream and see if that works
2. Try the fastPut method and see if that works

This module is really just a thin wrapper around the SSH2 and SFTP_STREAMS libraries. It is these libraries which do the heavy lifting.  It is possible there is something buggy in how the stream signals are being handled, though I would expect this would lead to errors on all platforms if that were the case. 

My only other suggestion would be to bypass ssh2-sftp-client and directly use the ssh2 and sftp-streams libraries. If you can reproduce the error, you can log an issue with those libraries.  If it works using the low level libs, then we would need to look at how this module is structured to see if there is some issue with stream signal handling unique to Windows environments.

**Comment by theophilusx** [comment]
The ECONNRESET error is generated when the remote end abruptly drops the connection. I've found this can happen in some environments if you are attempting to make too many connections too quickly. For example, for one test environment, I found it was necessary to add a delay between the test suites because after 12 connections, the remote server would drop any additional connections. This was due to a firewall setting/rule on the remote site. 

The root of the issue was that although the tests were closing the connections, under TCP, the connection is not dropped immediately. The socket goes into TIME_WAIT to allow any data still in the pipe to be cleared before the socket is completely released. The tests were running so fast that we ended up with enough connections in TIME_WAIT state to trip the rate limiting rule in the firewall. Adding a delay between tests solved this problem.

Although I don't think this is your problem, it is something to keep in mind. Based on the changes made in the  current version (v4.x), I suspect this issue may have been resolved. There was a bug in the end() method which could have triggered the error you have seen. This needs to be tested with the new version to see if the issue still exists.
